Add support for setting a disk offset when hibernating

superm1 opened the pull request:
This allows swapfiles to be more usable as users will no longer need to set the offset on their kernel command line manually.

Use this API in systemd when hibernating as well.

**NOTE:** The associated patch hasn't been merged into the Linux kernel yet.  This shouldn't be merged into systemd until it has (in case `/sys/power/disk_offset` is changed to a different name etc based on feedback).
_The kernel patch is submitted here if you would like to review:_
https://www.spinics.net/lists/linux-acpi/msg81033.html

This PR is submitted right now to facilitate parallel review of the kernel and userspace portions of this.

superm1 commented:
@ColinIanKing I borrowed your GPL extent reading code for this, I suppose I should update the copyright also and wanted to check to you should it be  `Copyright (C) 2010-2017 Canonical` ?

ColinIanKing commented:
@superm1 sure, that copyright is fine by me.

poettering commented on src/sleep/sleep.c:
```diff
@@ -38,8 +39,192 @@
+                free(result_fiemap);
+
+        if (fiemap)
+                free(fiemap);
```
usually, we rely on `_cleanup_free_` for making such clean-up paths automatic. please use that here too (moreover free() is actually fine with NULL parameters, hence you could drop the explicit checks anyway)

poettering commented on src/sleep/sleep.c:
```diff
@@ -38,8 +39,192 @@
+         *  of the last extent
+         */
+        while (fiemap_start < fiemap_length) {
+                memset(fiemap, 0, sizeof(struct fiemap));
```
zero(fiemap)

poettering commented on src/sleep/sleep.c:
```diff
@@ -38,8 +39,192 @@
+                if (ioctl(fd, FS_IOC_FIEMAP, fiemap) < 0) {
+                        fprintf(stderr, "fiemap ioctl() FS_IOC_FIEMAP failed, errno=%d (%s)\n",
+                                errno, strerror(errno));
+                        goto fail_cleanup;
```
as above, let the caller log, propagate -errno here

poettering commented on src/sleep/sleep.c:
```diff
@@ -38,8 +39,192 @@
+
+                /* Resize fiemap to allow us to read in the extents */
+                fm_tmp = realloc(fiemap,
+                                 sizeof(struct fiemap) + extents_size);
```
I think it would be nice to use GREEDY_REALLOC() here, like we usually do, to exponentially grow the buffer

poettering commented on src/sleep/sleep.c:
```diff
@@ -38,8 +39,192 @@
+                }
+                fiemap = fm_tmp;
+
+                memset(fiemap->fm_extents, 0, extents_size);
```
memzero()

poettering commented on src/sleep/sleep.c:
```diff
@@ -38,8 +39,192 @@
+
+                /* Highly unlikely that it is zero */
+                if (fiemap->fm_mapped_extents) {
+                        const __u32 i = fiemap->fm_mapped_extents - 1;
```
const? that sounds like a C++ism here...

also, please stick to C99 types, i.e. uint32_t

poettering commented on src/sleep/sleep.c:
```diff
@@ -38,8 +39,192 @@
+                return 0;
+
+        /* if it's a swap partition, we don't append offset */
+        if (strcmp(type, "partition") == 0)
```
We generally prefer "streq()" for comparing strings in place of "strcmp() == 0"

poettering commented on src/sleep/sleep.c:
```diff
@@ -38,8 +39,192 @@
+        /* if it's a swap partition, we don't append offset */
+        if (strcmp(type, "partition") == 0)
+                return write_string_file("/sys/power/disk_offset", device, 0);
+        else if(strcmp(type, "file") != 0) {
```
as above... also whitespace missing

poettering commented on src/sleep/sleep.c:
```diff
@@ -38,8 +39,192 @@
+        if (strcmp(type, "partition") == 0)
+                return write_string_file("/sys/power/disk_offset", device, 0);
+        else if(strcmp(type, "file") != 0) {
+                log_error_errno(r, "Unrecognized swap type '%s' %m", type);
```
"r" does not contain any appropriate error here, it shouldn't be included in this log line (that said, i doubt we should log at all here...)

poettering commented on src/sleep/sleep.c:
```diff
@@ -38,8 +39,192 @@
+                log_debug("Unable to stat %s", device);
+                return r;
+        }
+        fd = open(device, O_RDONLY);
```
please always pass O_CLOEXEC to open(), unless you have a really good reason not to.

poettering commented on src/sleep/sleep.c:
```diff
@@ -38,8 +39,192 @@
+        fd = open(device, O_RDONLY);
+        if (fd < 0) {
+                log_debug_errno(fd, "Unable to open '%s' %m", device);
+                return -EBADF;
```
please propagate the right error here, i.e. "return -errno".

poettering commented on src/sleep/sleep.c:
```diff
@@ -38,8 +39,192 @@
+        }
+        fd = open(device, O_RDONLY);
+        if (fd < 0) {
+                log_debug_errno(fd, "Unable to open '%s' %m", device);
```
fd doesn't contain any error here, errno does, hence the first argument should be "errno", not "fd"

poettering commented on src/sleep/sleep.c:
```diff
@@ -38,8 +39,192 @@
+                log_debug("Not enough memory for %d string", r);
+                return -ENOMEM;
+        }
+        snprintf(write, r, "%lx;%" PRIu64, stb.st_dev, offset);
```
please use glibc's asprintf() for this. It appears this is a manual reimplementation of that routine

superm1 commented:
Thank you for your comments, upstream is still in discussion on the associated kernel patch.  Once that discussion is complete I'll make adjustments to this series for your comments.

superm1 commented:
@poettering I've adjusted the PR for all your comments as well as for the comments that have occurred in the kernel.  In the kernel the desire is to keep writing the "disk" in `/sys/power/resume` and writing the offset in a separate file `/sys/power/resume_offset`.

So I've resubmitted the kernel patch for that and adjusted this PR as well.

poettering commented on src/shared/sleep-config.c:
```diff
@@ -224,13 +221,22 @@ static int hibernation_partition_size(size_t *size, size_t *used) {
-                *used = used_field;
+                if (device) {
+                        *device = dev_field;
+                        dev_field = NULL;
```
please use the TAKE_PTR() macro for the above (It's a very recent addition to systemd git, but makes passing of ownership much nicer and shorter)

poettering commented on src/shared/sleep-config.c:
```diff
@@ -224,13 +221,22 @@ static int hibernation_partition_size(size_t *size, size_t *used) {
+                }
+                if (type) {
+                        *type = type_field;
+                        type_field = NULL;
```
same here

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,161 @@
+
+        if (fstat(fd, &statinfo) != 0) {
+                log_debug_errno(errno, "Cannot determine file size (%s)\n %m",
+                                strerror(errno));
```
%m already outputs the error string, no need to use strerror() anymore.

Please change this to:

```c
if (fstst(fd, &statinfo) < 0) 
        return log_debug_errno(errno, "Cannot determine file size: %m");
```

This takes benefit of the fact that log_debug_errno() returns the negative error you pass into it anyway

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,161 @@
+                                strerror(errno));
+                return -errno;
+        }
+        fiemap_length = statinfo.st_size;
```
a safety check S_ISREG() would be wise here I think

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,161 @@
+
+                fiemap->fm_start = fiemap_start;
+                fiemap->fm_length = fiemap_length;
+                fiemap->fm_flags = FIEMAP_FLAG_SYNC;
```
instead of zeroing out fiemap, and then filling in fields, it#s much nicer to read and better for the compiler to just use C99 structure initializers:

```c
*fiemap = (struct fiemap) = {
        .fm_start = fiemap_start,
        .fm_length = fiemap_length,
        .fm_flags = FIEMAP_FLAG_SYNC,
};
```

In that case the compiler will zero-reset all other fields and any padding, fully automatically.

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,161 @@
+
+                /* Resize fiemap to allow us to read in the extents */
+                if (!GREEDY_REALLOC(fiemap, fiemap_size,
+                                    sizeof(struct fiemap) + extents_size))
```
`offsetof(struct fiemap, fm_extents) + extents_size`...

using offsetoff instead of sizeof() doesn't change much here on the compiler level, since the fiemap structure has no whitespace, but it's still nicer to read I think, and doesn't imply that the structure is set up in some specific way.

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,161 @@
+                result_extents += fiemap->fm_mapped_extents;
+
+                /* Highly unlikely that it is zero */
+                if (fiemap->fm_mapped_extents) {
```
we try to avoid on relying on C's downgrade-to-bool feature when checking numeric variables. We do that only for stuff that is boolean or boolean-like (or for pointers). Hence, please write this as `if (fiemap->fm_mapped_extens > 0)` or so

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,161 @@
+
+        result_fiemap->fm_mapped_extents = result_extents;
+        *ret = result_fiemap;
+        result_fiemap = NULL;
```
TAKE_PTR

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,161 @@
+        r = stat(device, &stb);
+        if (r < 0) {
+                log_debug("Unable to stat %s", device);
+                return r;
```
```c
return log_debug_errno(errno, "Unable to stat '%s': %m", device)
```

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,161 @@
+        /* Only available in 4.17+ */
+        r = access("/sys/power/resume_offset", W_OK);
+        if (r < 0) {
+                log_debug_errno(r, "/sys/power/resume_offset unavailable: %m");
```
the first parameter of log_debug_errno() should be the error code, but access() returns that in `errno`, not in its return value, hence this should be changed to: "log_debug_errno(errno, …"

poettering replied:
also, I'd prefer a more strict check, i.e. we should assume that resume_offset is unavailable only when access(F_OK) returns ENOENT. In all other cases we should consider this a proper error to propagate, since something else is wrong...

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,161 @@
 
 static char* arg_verb = NULL;
 
+static int read_fiemap(int fd, struct fiemap **ret) {
```
in systemd we try to make sure that functions either log about none of the errors above LOG_DEBUG level (in which case the caller is supposed to do that), or about all errors. The former is the intended scheme for "library-like" functions, and the latter for "main-like" functions, but the line is blurry when a function is clearly one or the other. 

This function logs about some errors at LOG_ERR (specifically: OOM), and about others it just does debug logging or none at all. Please clean this up: either log about all errors here, or about none/downgrade all logs to LOG_DEBUG.

Given that this kinda feels like a "library-like" API I figure it would be better to not log above LOG_DEBUG in this function.

Strictly following logging rules like this ensures that errors are logged about exactly once, as logging about them multiple times or not at all is not helpful to the user.

superm1 replied:
OK I think I've got it how you intend now (more library like).  Please let me know if I still have something off.

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,161 @@
+
+        r = find_hibernate_location(&device, &type, NULL, NULL);
+        if (r < 0)
+                return 0;
```
hmm, you are ignoring all errors from find_hibernate_location() here, mostly silently. That's not a good idea, is it?

superm1 replied:
Yeah, this was from before I had made the change to only call write_disk_offset guarded by:
```
if(!strv_isempty(modes)) {
```

It made sense because the code was also called in suspend, but now that it's only called in hibernate, those errors should be raised.

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,161 @@
+                log_debug("Unable to stat %s", device);
+                return r;
+        }
+        fd = open(device, O_RDONLY | O_CLOEXEC);
```
this smells like TTOCTTOU. Please open() first, then fstat() the fd.

poettering replied:
Also, please add O_NONBLOCK here, so that we don't end up blocking on weird block devices (think: some weird stuff like floppys which block as long as nothing is in the drive)

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,161 @@
+        fd = open(device, O_RDONLY | O_CLOEXEC);
+        if (fd < 0) {
+                log_debug_errno(errno, "Unable to open '%s' %m", device);
+                return -errno;
```
as above: `return log_debug_errno(errno, "Unable to open '%s': %m", device)`

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,161 @@
+        close(fd);
+        if (r < 0) {
+                log_debug("Unable to read extent map for '%s' (%d) %m",
+                          device, r);
```
`return log_debug_errno(r, …)`

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,161 @@
+                log_debug("No extents found in '%s'", device);
+                return -EINVAL;
+        }
+        offset = fiemap->fm_extents[0].fe_physical / getpagesize();
```
please use our caching wrapper page_size()

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,161 @@
+        r = write_string_file("/sys/power/resume_offset", offset_str, 0);
+        if (r < 0) {
+                log_debug("Failed to write offset: '%d' %m", r);
+                return r;
```
return log_debug_errno(r, "Failed to write...)

poettering commented on src/sleep/sleep.c:
```diff
@@ -110,9 +268,14 @@ static int execute(char **modes, char **states) {
-        r = write_mode(modes);
-        if (r < 0)
-                return r;
+        if(!strv_isempty(modes)) {
```
space between "if" and "(", please

poettering commented on src/sleep/sleep.c:
```diff
@@ -23,6 +24,8 @@
 #include <getopt.h>
 #include <stdio.h>
+#include <linux/fs.h>
+#include <linux/fiemap.h>
```
btw, we try to keep `#include` lists sorted alphabetically, see CODING_STYLE...

superm1 replied:
Thanks.  Do you want Christmas tree + alphabetical, or purely alphabetical?
Eg:
```
 #include <errno.h>
 #include <getopt.h>
 #include <stdio.h>
+#include <linux/fiemap.h>
+#include <linux/fs.h>
```
or 
```
 #include <errno.h>
 #include <getopt.h>
+#include <linux/fiemap.h>
+#include <linux/fs.h>
 #include <stdio.h>
```

poettering replied:
purely alphabetical, see doc/CODING_STYLE

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,160 @@
+
+        /* Only available in 4.17+ */
+        r = access("/sys/power/resume_offset", F_OK);
+        if (r == ENOENT)
```
access() returns errors in `errno`, not in its return value, you hence need to check:

```c
if (access("…", F_OK) < 0) {
        if (errno == ENOENT)
                return 0;
        return -errno;
}
```
or something like that

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,160 @@
+        fd = open(device, O_RDONLY | O_CLOEXEC | O_NONBLOCK);
+        if (fd < 0)
+                return log_debug_errno(errno, "Unable to open '%s': %m", device);
+        r = stat(device, &stb);
```
`fstat(fd, &stb);`

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,160 @@
+                return log_debug_errno(errno, "Unable to open '%s': %m", device);
+        r = stat(device, &stb);
+        if (r < 0) {
+                close(fd);
```
instead of closing here manually, please just declare `fd` as `_cleanup_close_ int fd = -1;` so that the compiler generates the clean-up code for us automatically.

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,160 @@
+                return log_debug_errno(errno, "Unable to stat %s: %m", device);
+        }
+        r = read_fiemap(fd, &fiemap);
+        close(fd);
```
with the suggested `_cleanup_close_` declaration you can then also drop this line

poettering commented on src/sleep/sleep.c:
```diff
@@ -110,9 +267,14 @@ static int execute(char **modes, char **states) {
-        if (r < 0)
-                return r;
+        if (!strv_isempty(modes)) {
+                r = write_disk_offset();
```
so, currently if something fails in write_disk_offset() we don't log at all above debug logging. The idea is that either the functions you call log something actionable themselves, or the caller does it. Now you downgraded all logging within the function to LOG_DEBUG (or none at all), but that means you really should log here when calling it, i.e. change this:

```c
r = write_disk_offset();
if (r < 0)
        log_error_errno(r, "Failed to write hibernation disk offset: %m");
```

The rule is that every error should be logged exactly once above LOG_DEBUG level, and that means both *at least* once, and *at most* once...

poettering replied:
(and yupp, some of our current codebase doesn't get right, but for new code we try to strictly follow this)

superm1 replied:
OK.  I think i'll do a second commit to move the logging in `write_mode` to be match this behavior as well then.

superm1 commented:
The kernel has merged the patch to add a new sysfs file resume_offset.

superm1 commented:
@poettering all your previous comments have been accounted for. Can you please give another review when you get a chance?

poettering commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,161 @@
+        r = fstat(fd, &stb);
+        if (r < 0) {
+                return log_debug_errno(errno, "Unable to stat %s: %m", device);
+        }
```
nitpick: no {} around single-line if blocks, see CODING_STYLE

poettering commented on src/sleep/sleep.c:
```diff
@@ -287,7 +284,7 @@ static int execute(char **modes, char **states) {
         r = write_state(&f, states);
         if (r < 0)
-                return r;
+                return log_error_errno(errno, "Failed to write /sys/power/state: %m");
```
write_state() returns the error as return value, hence the first argument to log_error_errno() should be `r`, not `errno`

poettering commented:
looks very good. found one last issue though (and one nitpick). Please fix and it's good for merging.

superm1 commented:
Thanks, cleaned up those last two and force pushed just now.

poettering commented:
excellent! thanks!

keszybz commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,160 @@
+        if (fstat(fd, &statinfo) < 0)
+                return log_debug_errno(errno, "Cannot determine file size: %m");
+        if (!S_ISREG(statinfo.st_mode))
+                return -EBADF;
```
This should be ENOTTY.

keszybz commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,160 @@
+                if (fiemap->fm_mapped_extents == 0)
+                        break;
+
+                /* Result fiemap have to hold all the extents for the hole file */
```
ha**s** to hold ... for the **w**hole file

keszybz commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,160 @@
+
+                /* Read in the extents */
+                extents_size = sizeof(struct fiemap_extent) *
+                                      (fiemap->fm_mapped_extents);
```
Please drop the parens around `fiemap->fm_mapped_extents` and make this one line, it's confusing when written like it is now.

keszybz commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,160 @@
+
+                /* Resize fiemap to allow us to read in the extents */
+                if (!GREEDY_REALLOC(fiemap, fiemap_size,
+                                    offsetof(struct fiemap, fm_extents) + extents_size))
```
GREEDY_REALLOC takes the number of items as the last argument, not bytes.

keszybz commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,160 @@
+                                    offsetof(struct fiemap, fm_extents) + extents_size))
+                        return -ENOMEM;
+
+                memzero(fiemap->fm_extents, extents_size);
```
Use GREEDY_REALLOC0 and then this memzero is not needed.

keszybz commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,160 @@
+
+                extents_size = sizeof(struct fiemap_extent) *
+                                      (result_extents +
+                                       fiemap->fm_mapped_extents);
```
Please just make this one line.

keszybz commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,160 @@
+
+                /* Resize result_fiemap to allow us to read in the extents */
+                if (!GREEDY_REALLOC(result_fiemap, result_fiemap_size,
+                                    offsetof(struct fiemap, fm_extents) + extents_size))
```
Here too, GREEDY_REALLOC takes number of items as the last arg.

keszybz commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,160 @@
+
+        r = find_hibernate_location(&device, &type, NULL, NULL);
+        if (r < 0)
+                return log_debug_errno(r, "Unable to find hibernation location.");
```
"Unable to find hibernation location: %m"

keszybz commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,160 @@
+        return 0;
+}
+
+static int write_disk_offset(void) {
```
Can we call this `write_hibernate_location_info` or something like that, since it doesn't just do offset.

keszybz commented on src/sleep/sleep.c:
```diff
@@ -40,6 +43,160 @@
+
+                memcpy(result_fiemap->fm_extents + result_extents,
+                       fiemap->fm_extents,
+                       sizeof(struct fiemap_extent) *
```
It'd be more legible if the last arg wasn't split on two lines.

superm1 commented:
@keszybz okay I've added a test for this and adjusted for your other comments. Greedy realloc doesn't fit this perfectly since it's being resized as the size of the structure plus some different size of the extent map. Since this was not like an array with same size of each element the 4th argument was throwing it all off. So I've made all the arguments relative to the size of the structure fiemap instead.

keszybz commented on src/test/test-sleep.c:
```diff
@@ -18,13 +18,41 @@
+        log_info("\t number of mapped extents: %u", fiemap->fm_mapped_extents);
+        log_info("\t extent count: %u", fiemap->fm_extent_count);
+        log_info("\t first extent location: %llu",
+                 fiemap->fm_extents[0].fe_physical / page_size());
```
Thanks for this.

Hmm, here I see:
```
extent map information for /home/zbyszek/src/systemd/build/systemd-sleep:
	 start: 94545859689904
	 length: 8299960843193248865
	 flags: 1702130553
	 number of mapped extents: 1
	 extent count: 1885693292
	 first extent location: 12099829
```
Are those large numbers exptected?
(This is on ext4.)

superm1 commented:
You know I was only really interested in the last one and wasn't paying attention to those. I forgot they need to be copied at the end to be valid. Fixed up.

Thanks for catching that!

keszybz commented on src/test/test-sleep.c:
```diff
@@ -18,13 +18,41 @@
+        _cleanup_free_ struct fiemap *fiemap = NULL;
+        _cleanup_close_ int fd = -1;
+        int r;
+        const char *device = ABS_BUILD_DIR "/systemd-sleep";
```
This would work most of the time, but we also allow tests to be installed. So the test binary should work even if the build directory is removed. Hence getting access to the files in the build tree is rather complicated, see get_testdata_dir() if interested. Fortunately, here we can just use argv[0], it'll work just as well.

I'll push a follow-up patch that allows testing on arbitrary files and uses argv[0]. Please squash it in in the next round.

superm1 commented:
One more change just now (in case not obvious with the force push).  I noticed RPM CI failed because `test-sleep` now needs root, so I'm checking if root in `test-sleep` for CI that doesn't run as root.

keszybz commented:
Oh, github doesn't allow me to upload a .patch file now as an attachment. For heavens' sake.

```diff
From 6345cf24450608fb28309c0119f92f3b6fae62fa Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Zbigniew=20J=C4=99drzejewski-Szmek?= <zbyszek@in.waw.pl>
Date: Mon, 9 Apr 2018 14:55:43 +0200
Subject: [PATCH] test-sleep: add support for testing fiemap on arbitrary files

Also, let's not use ABS_BUILD_DIR, because that is incompatible with installed
tests.
---
 src/test/test-sleep.c | 28 +++++++++++++++++++---------
 1 file changed, 19 insertions(+), 9 deletions(-)

diff --git a/src/test/test-sleep.c b/src/test/test-sleep.c
index 86bf7b6196..1ee2eaa2c5 100644
--- a/src/test/test-sleep.c
+++ b/src/test/test-sleep.c
@@ -27,21 +27,19 @@
 #include "strv.h"
 #include "util.h"
 
-static int test_fiemap(void) {
+static int test_fiemap(const char *path) {
         _cleanup_free_ struct fiemap *fiemap = NULL;
         _cleanup_close_ int fd = -1;
         int r;
-        const char *device = ABS_BUILD_DIR "/systemd-sleep";
 
-        fd = open(device, O_RDONLY | O_CLOEXEC | O_NONBLOCK);
+        fd = open(path, O_RDONLY | O_CLOEXEC | O_NONBLOCK);
         if (fd < 0)
-                return log_error_errno(errno, "failed to open %s: %m", device);
+                return log_error_errno(errno, "failed to open %s: %m", path);
         log_info("fd: %d", fd);
         r = read_fiemap(fd, &fiemap);
         if (r < 0)
-                return log_error_errno(r, "Unable to read extent map for '%s': %m",
-                                       device);
-        log_info("extent map information for %s:", device);
+                return log_error_errno(r, "Unable to read extent map for '%s': %m", path);
+        log_info("extent map information for %s:", path);
         log_info("\t start: %llu", fiemap->fm_start);
         log_info("\t length: %llu", fiemap->fm_length);
         log_info("\t flags: %u", fiemap->fm_flags);
@@ -49,8 +47,8 @@ static int test_fiemap(void) {
         log_info("\t extent count: %u", fiemap->fm_extent_count);
         log_info("\t first extent location: %llu",
                  fiemap->fm_extents[0].fe_physical / page_size());
+
         return 0;
-
 }
 
 static void test_sleep(void) {
@@ -80,6 +78,8 @@ static void test_sleep(void) {
 }
 
 int main(int argc, char* argv[]) {
+        int i, r = 0, k;
+
         log_parse_environment();
         log_open();
 
@@ -87,5 +87,15 @@ int main(int argc, char* argv[]) {
                 log_warning("This program is unlikely to work for unprivileged users");
 
         test_sleep();
-        return test_fiemap();
+
+        if (argc <= 1)
+                assert_se(test_fiemap(argv[0]) == 0);
+        else
+                for (i = 1; i < argc; i++) {
+                        k = test_fiemap(argv[i]);
+                        if (r == 0)
+                                r = k;
+                }
+
+        return r;
 }
-- 
```

>  noticed RPM CI failed because test-sleep now needs root

Why? The added test does not require root privs.

superm1 commented:
> Why? The added test does not require root privs.
Hmm, something is causing the ioctl call to fail in the RPM build then.  I'll drop the root check and investigate when I push again to account for your other comments.

superm1 commented:
@keszybz Alright I've cleaned up the problems around files not on a filesystem and squashed your patch.

keszybz commented:
Looks nice.

medhefgo commented:
If I am not mistaken, the kernel still needs to be told about this resume disk+offset on next boot or there will be a normal boot, right? It would be great if you could integrate this with systemd-boot by having this also written to the ESP / boot partition so the bootloader can pick it up and append to the kernel command line.

On a related note, I'd love if one could reserve swap space for hibernation only. That way one could be guaranteed that hibernation will never abort due to too little swap space, while also not having to extremely over-provision the swap space. Would be nice if you could look into that....
